src/oracle/middleware.js

```diff
@@ -0,0 +1,761 @@
+
+    res.status(200).send({ success: true })
+  });
+ 
+  const listener = app.listen(PORT, '0.0.0.0', () => {
+    console.log(`External adapter listening on port`, PORT);
+  });
+
+  listener.on('error', err => { console.log("Bridge found error:", err) })
+}
+
+/** @param {import('../lib/psm.js').BridgeAction} bridgeAction */
+const outputAction = bridgeAction => {
+  const capData = marshaller.serialize(bridgeAction);
+  var data = JSON.stringify(capData)
+  return data
+};
+
+/**
+ * Function to check if currently in submission
+ * @param {*} feed feed to check for
+ * @returns whether last submission was made in less than SEND_CHECK_INTERVAL seconds
+ */
+const checkIfInSubmission = async (feed) => {
+
+  //get last submission time
+  let query = await queryTable("jobs", ["last_submission_time"], feed)
+  //get seconds since last price submission
+  let timePassedSinceSubmission = (Date.now()/1000) - 
+  query.last_submission_time
+  return timePassedSinceSubmission < Number(SEND_CHECK_INTERVAL)
+}
+
+/**
+  * Function to push price on chain to the smart wallet
+  * @param {*} price price to push
+  * @param {*} feed feed to push price to
+  * @param {*} round round to push result to
+  * @param {*} from account to push from
+  * @returns whether successful
+  */
+const pushPrice = async (price, feed, round, from) => {
+
+  //create an offerId with the Date number
+  var offerId = Date.now()
+
+  //get offers
+  let offers = await getOraclesInvitations()
+  //get previous offer for feed
+  let previousOffer = offers[feed]
+
+  //create an offer
+  const offer = {
+    id: Number(offerId),
+    invitationSpec: {
+      source: 'continuing',
+      previousOffer: Number(previousOffer),
+      invitationMakerName: 'PushPrice',
+      invitationArgs: harden([{ unitPrice: BigInt(price), roundId: round }]),
+    },
+    proposal: {},
+  };
+
+  //create keyring
+  var keyring = {
+    "home": "",
+    "backend": "test"
+  }
+
+  //check if submitted for round
+  let submitted = await checkSubmissionForRound(from, previousOffer, round)
+
+  //check if in submission
+  let inSubmission = await checkIfInSubmission(feed)
+
+  //loop retries
+  for (let i = 0; i < SUBMIT_RETRIES && !submitted && !inSubmission; i++) {
+
+    //query round
+    let latestRound = await queryRound(feed)
+    
+    //if latestRound is greater than round being pushed or submission to the round is already made, abort
+    if (latestRound.round_id > round || (latestRound.round_id == round && latestRound.submission_made)){
+      console.log("Price failed to be submitted for old round", round)
+      return false
+    }
+
+    console.log("Submitting price for round", round, "try", (i + 1))
+
+    offer.id = Number(Date.now())
+
+    //output action
+    var data = outputAction({
+      method: 'executeOffer',
+      // @ts-ignore
+      offer,
+    });
+
+    //change data to JSON
+    data = JSON.parse(data)
+
+    //execute
+    await execSwingsetTransaction(
```

Comment:
This command can easily fail (this is basically making a tx) but no way to determine a tx failed or was actually successful? In the below code we assume that it was successful but it can't be always the case

Reply:
This is fine because it will be retried if a submission is not found, also this function was taken from agoric-sdk and it does not return anything back.